tests/e2e/src/main.rs

```diff
@@ -0,0 +1,219 @@
+    let connection = create_connection(&config);
+    // NOTE : we wait here for a whole genesis session to pass
+    // session period is set to 5 blocks (see `run_consensus.sh`), plus one to be on the safe site
+    wait_for_block(connection, 6)
+}
+
+fn test_token_transfer(config: Config) -> anyhow::Result<()> {
+    let from: sr25519::Pair = config
+        .accounts
+        .get(0)
+        .expect("No accounts passed")
+        .to_owned();
+    let to = AccountId::from(
+        config
+            .accounts
+            .get(1)
+            .expect("Pass at least two accounts")
+            .public(),
+    );
+
+    let connection = create_connection(&config).set_signer(from);
+
+    let balance_before = connection
+        .get_account_data(&to.clone())?
+        .expect("Could not get account data")
+        .free;
+
+    println!("[+] Account {} balance before tx: {}", to, balance_before);
+
+    let transfer_value = 1000u128;
+
+    let tx: UncheckedExtrinsicV4<_> = compose_extrinsic!(
+        connection,
+        "Balances",
+        "transfer",
+        GenericAddress::Id(to.clone()),
+        Compact(transfer_value)
+    );
+
+    // send and watch extrinsic until InBlock
+    let tx_hash = connection
+        .send_extrinsic(tx.hex_encode(), XtStatus::InBlock)?
+        .expect("Could not get tx hash");
+
+    println!("[+] Transaction hash: {}", tx_hash);
+
+    let balance_after = connection
+        .get_account_data(&to.clone())?
+        .expect("Could not get account data")
+        .free;
+
+    println!("[+] Account {} balance after tx: {}", to, balance_after);
+
+    assert!(
+        balance_before + transfer_value == balance_after,
+        "before = {}, after = {}, tx = {}",
+        balance_before,
+        balance_after,
+        transfer_value
+    );
+
+    Ok(())
+}
+
+fn test_change_validators(config: Config) -> anyhow::Result<()> {
+    let sudo: sr25519::Pair = config.clone().sudo;
+    let connection = create_connection(&config).set_signer(sudo);
+
+    let validators_before: Vec<AccountId> = connection
+        .get_storage_value("Session", "Validators", None)?
+        .unwrap();
+
+    info!("[+] Validators before tx: {:#?}", validators_before);
+
+    let mut new_validators: Vec<AccountId> = config
+        .accounts
+        .into_iter()
+        .map(|pair| pair.public().into())
+        .collect();
+
+    new_validators.push(
+        AccountId::from_ss58check("5EHkv1FCd4jeQmVrbYhrETL1EAr8NJxNbukDRT4FaYWbjW8f").unwrap(),
+    );
+
+    info!("[+] New validators {:#?}", new_validators);
+
+    // what session is this?
+
+    let session_number: u32 = connection
+        .get_storage_value("Session", "CurrentIndex", None)
+        .unwrap()
+        .unwrap();
+
+    info!("[+] Session number {:?}", session_number);
+
+    let call = compose_call!(
+        connection.metadata,
+        "Aleph",
+        "change_validators",
+        new_validators.clone(),
+        session_number + 1
```

Comment:
`+1` is risky, right? It may happen that in between getting the session number and submitting the `change_validators` xt, a block from new session was emitted, then the change won't be enacted, right?

Reply:
Done